pkg/controllers/reconciler.go

```diff
@@ -17,28 +17,42 @@ package controllers
 
 import (
 	"context"
+	"path/filepath"
+	"strings"
 	"sync"
 
 	"github.com/go-logr/logr"
+	"github.com/google/go-cmp/cmp"
 	apps "k8s.io/api/apps/v1"
 	v1 "k8s.io/api/core/v1"
 	rbacv1 "k8s.io/api/rbac/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/diff"
+	"k8s.io/apimachinery/pkg/util/version"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/log"
 )
 
 var (
-	bKeeper = &bookKeeper{}
+	once             Once
+	UpgradePlugins   bool
+	UpgradeToVersion string
+	bKeeper          = &bookKeeper{}
 )
 
+const ImageMinVersion = "0.23.0"
+
 func init() {
 	bKeeper.pluginCounter = make(map[string]int)
 }
 
+type Once struct {
```

Comment:
nit: I'd rather hide this type and the variables like `UpgradePlugins`, `UpgradeToVersion` from public interface. BTW, not sure this type helps readability. Otherwise looks very good.

Reply:
@rojkov @mythi 
> I'd rather hide this type [...] BTW, not sure this type helps readability

Thanks for the input and fully agree. The problem here is linter (which ATM simply counts "if"s, and Reconcile() is already at the limit), so this trickery is just tame lint.

I had a few experiments to reduce cyclo whithout harming readability of Reconcile(), but wasn't too pleased with results (brainstorming it and will return with some solution after this PR).

Let me check if these simplest fixes would do.